File: mobile/welcome-screen/android-preview-project/app/src/main/java/com/showup/observability/Crashes.kt
package com.showup.observability

import android.content.Context
import io.sentry.SentryEvent
import io.sentry.SentryOptions
import io.sentry.android.core.SentryAndroid
import io.sentry.protocol.User

/**
 * Crash and error reporting, off unless it is switched on and given a destination.
 *
 * WHY THE SWITCH, AND WHY OFF IS THE DEFAULT
 *
 * The same shape as the backend's `initSentry` (Epic 16): with `SENTRY_ENABLED` false, or with no
 * DSN, nothing is initialised and nothing is ever sent -- so the app builds and runs with no
 * account, no credentials and no network calls to a third party. Anyone can clone this repository
 * and build it without asking for a key.
 *
 * WHAT IS DELIBERATELY NOT COLLECTED
 *
 * A crash reporter is a pipe out of the device, and this is a dating app: the data within reach of
 * a stack trace includes phone numbers, one-time codes, coordinates and message text. So:
 *
 *   - `isSendDefaultPii = false`, so the SDK never attaches IP address or device identifiers of
 *     its own accord.
 *   - `beforeSend` walks every event and redacts any field whose NAME is prohibited, using the same
 *     list as the backend. Names, not values -- a value-based check cannot recognise a phone number
 *     it has not seen the format of.
 *   - breadcrumbs are dropped entirely. They are the most useful feature here and also the most
 *     dangerous: they capture UI interactions and network URLs automatically, which on the phone
 *     screen means the number being typed. Reinstate them only with an explicit allowlist.
 *   - session tracking is off, because it exists to measure release health and needs an installation
 *     identifier to do it.
 *
 * Auto-init is disabled in the manifest, so this function is the only way the SDK ever starts.
 */
object Crashes {

    /**
     * Starts reporting if configured, and returns whether it did.
     *
     * A thin wrapper. The gate and the whole configuration live in [configure], which needs no
     * Android at all -- so the rules that actually matter are unit-tested without Robolectric, an
     * emulator, or a pinned SDK level.
     */
    fun start(
        context: Context,
        enabled: Boolean,
        dsn: String,
        environment: String,
        release: String,
    ): Boolean {
        // Checked here as well as in configure, because the SDK must not be touched at all when
        // reporting is off -- SentryAndroid.init installs integrations before the options lambda
        // has any say.
        if (!enabled || dsn.isBlank()) return false

        SentryAndroid.init(context) { options ->
            configure(options, dsn, environment, release)
        }
        return true
    }

    /**
     * Applies every setting, and returns whether reporting should happen.
     *
     * Separate from [start] and free of Android types on purpose: this is the part with the rules
     * in it, and a rule nobody can test cheaply is a rule that quietly stops being true.
     */
    internal fun configure(
        options: SentryOptions,
        dsn: String,
        environment: String,
        release: String,
    ): Boolean {
        if (dsn.isBlank()) return false

        options.dsn = dsn
        options.environment = environment
        options.release = release

        // Never let the SDK attach personal information of its own accord.
        options.isSendDefaultPii = false

        // Release health needs an installation identifier to attribute sessions. Not worth a
        // persistent identifier on this app.
        options.isEnableAutoSessionTracking = false

        // See the note above: breadcrumbs auto-capture typed UI and request URLs.
        options.maxBreadcrumbs = 0
        options.isEnableUserInteractionBreadcrumbs = false

        options.beforeSend = SentryOptions.BeforeSendCallback { event, _ -> scrub(event) }
        return true
    }

    /**
     * Redacts prohibited fields from an event, in place.
     *
     * Returns the event rather than null: dropping the whole report would lose the crash as well as
     * the personal data, and the crash is the reason any of this exists.
     */
    internal fun scrub(event: SentryEvent): SentryEvent {
        // A user object here would carry an id, an email or an IP. None of it is needed to fix a
        // crash, and all of it is personal data leaving the device.
        event.user = null as User?

        event.extras = event.extras?.mapValues { (key, value) ->
            if (ProhibitedFields.isProhibited(key)) REDACTED else value
        }

        event.tags = event.tags?.mapValues { (key, value) ->
            if (ProhibitedFields.isProhibited(key)) REDACTED else value
        }

        // Breadcrumbs are configured off, but an SDK integration can still add one before this
        // runs, so they are cleared here too rather than trusted to stay empty.
        event.breadcrumbs = emptyList()

        return event
    }

    internal const val REDACTED = "[redacted]"
}

File: mobile/welcome-screen/ios-app/ShowUpWelcome/Crashes.swift
import Foundation
import Sentry

/// Crash and error reporting, off unless it is switched on and given a destination.
///
/// WHY THE SWITCH, AND WHY OFF IS THE DEFAULT
///
/// The same shape as the backend's `initSentry` (Epic 16) and Android's `Crashes`: with reporting
/// disabled, or with no DSN, nothing is started and nothing is ever sent — so the app builds and
/// runs with no account, no credentials and no calls to a third party. Anyone can clone this
/// repository and build it without asking for a key.
///
/// WHAT IS DELIBERATELY NOT COLLECTED
///
/// A crash reporter is a pipe off the device, and this is a dating app: the data within reach of a
/// stack trace includes phone numbers, one-time codes, coordinates and message text. So:
///
/// - `sendDefaultPii` is off, so the SDK never attaches an IP address or device identifiers of its
///   own accord.
/// - `beforeSend` redacts any field whose NAME is prohibited, using the same list as the backend and
///   Android. Names, not values — a value-based check cannot recognise a phone number in a format it
///   has not seen.
/// - breadcrumbs are dropped entirely. They are the most useful feature here and the most dangerous:
///   they capture UI interactions and network URLs automatically, which on the phone screen means
///   the number being typed. Reinstate only with an explicit allowlist.
/// - `enableAutoSessionTracking` is off, because release health needs an installation identifier.
/// - screenshots and view hierarchies are off. Both would capture a half-typed phone number, and
///   both are opt-in features that exist precisely because they are so useful — which is why they
///   have to be turned down explicitly rather than left at whatever the SDK defaults to next.
enum Crashes {

    /// Starts reporting if configured, and returns whether it did.
    ///
    /// - Parameter start: injectable so the gating is testable without launching the real SDK, the
    ///   same way the backend's version is.
    @discardableResult
    static func start(
        enabled: Bool,
        dsn: String,
        environment: String,
        release: String,
        start: (Options) -> Void = { SentrySDK.start(options: $0) }
    ) -> Bool {
        guard enabled, !dsn.trimmingCharacters(in: .whitespaces).isEmpty else { return false }

        let options = Options()
        configure(options, dsn: dsn, environment: environment, release: release)
        start(options)
        return true
    }

    /// Applies every setting.
    ///
    /// Separate from ``start(enabled:dsn:environment:release:start:)`` on purpose: this is the part
    /// with the rules in it, and a rule nobody can test cheaply is a rule that quietly stops being
    /// true.
    static func configure(
        _ options: Options,
        dsn: String,
        environment: String,
        release: String
    ) {
        options.dsn = dsn
        options.environment = environment
        options.releaseName = release

        options.sendDefaultPii = false
        options.enableAutoSessionTracking = false
        options.maxBreadcrumbs = 0
        options.enableUserInteractionTracing = false
        options.attachScreenshot = false
        options.attachViewHierarchy = false

        options.beforeSend = { event in Crashes.scrub(event) }
    }

    /// Redacts prohibited fields from an event.
    ///
    /// Returns the event rather than nil: dropping the whole report would lose the crash as well as
    /// the personal data, and the crash is the reason any of this exists.
    static func scrub(_ event: Event) -> Event {
        // A user object here would carry an id, an email or an IP. None of it is needed to fix a
        // crash, and all of it is personal data leaving the device.
        event.user = nil

        if let extra = event.extra {
            event.extra = extra.reduce(into: [String: Any]()) { out, pair in
                out[pair.key] = ProhibitedFields.isProhibited(pair.key) ? redacted : pair.value
            }
        }

        if let tags = event.tags {
            event.tags = tags.reduce(into: [String: String]()) { out, pair in
                out[pair.key] = ProhibitedFields.isProhibited(pair.key) ? redacted : pair.value
            }
        }

        // Breadcrumbs are configured off, but an SDK integration can still add one before this
        // runs, so they are cleared here too rather than trusted to stay empty.
        event.breadcrumbs = []

        return event
    }

    static let redacted = "[redacted]"
}

/// Whether this build reports crashes, and where to.
///
/// Both are compile-time constants and both are off in the repository, so no credential is
/// committed and the app builds for anyone. The release pipeline supplies them when there is one,
/// and the DSN must be an EU-region ingest host.
enum CrashReporting {
    static let enabled = false
    static let dsn = ""
}
